packages/block-editor/src/components/block-list/block.js

```diff
@@ -327,34 +330,135 @@ const applyWithDispatch = withDispatch( ( dispatch, ownProps, { select } ) => {
 		},
 		onInsertBlocksAfter( blocks ) {
 			const { clientId, rootClientId } = ownProps;
-			const { getBlockIndex } = select( blockEditorStore );
+			const { getBlockIndex } = registry.select( blockEditorStore );
 			const index = getBlockIndex( clientId );
 			insertBlocks( blocks, index + 1, rootClientId );
 		},
 		onMerge( forward ) {
 			const { clientId, rootClientId } = ownProps;
-			const { getPreviousBlockClientId, getNextBlockClientId, getBlock } =
-				select( blockEditorStore );
+			const {
+				getPreviousBlockClientId,
+				getNextBlockClientId,
+				getBlock,
+				getBlockAttributes,
+				getBlockName,
+				getBlockOrder,
+			} = registry.select( blockEditorStore );
 
+			// For `Delete` or forward merge, we should do the exact same thing
+			// as `Backspace`, but from the other block.
 			if ( forward ) {
+				if ( rootClientId ) {
+					const nextRootClientId =
+						getNextBlockClientId( rootClientId );
+
+					if ( nextRootClientId ) {
+						// If there is a block that follows with the same parent
+						// block name and the same attributes, merge the inner
+						// blocks.
+						if (
+							getBlockName( rootClientId ) ===
+							getBlockName( nextRootClientId )
+						) {
+							const rootAttributes =
+								getBlockAttributes( rootClientId );
+							const previousRootAttributes =
+								getBlockAttributes( nextRootClientId );
+
+							if (
+								Object.keys( rootAttributes ).every(
+									( key ) =>
+										rootAttributes[ key ] ===
+										previousRootAttributes[ key ]
+								)
```

Comment:
You raise an interesting question here. What led you to this check? For me, the first scenario that came to mind was merging between a UL and a OL. I first expected to just merge (i.e. skip this check), but on the other hand the way that Backspace unwraps the LI first means that the next Backspace will be very obvious (merge into the prev list, regardless of UL/OL).

Reply:
Yes, not sure if we should keep this check or not. This is generalised for all wrapper blocks. If two quotes have an empty or same citation, they can be merged, otherwise not. If two group blocks have the same background color, they can be merged, otherwise not. But you could also make the case that we should just merge them regardless of attributes, and maybe keep the attributes of the currently selected block.

Reply:
> If two quotes have an empty or same citation, they can be merged, otherwise not. If two group blocks have the same background color, they can be merged, otherwise not. 

At the risk of overpinging, @jasmussen might be someone who cares about these details like we do. :)

Reply:
I have this feeling — and correct me if I'm off beat here — that there's already a great deal of complexity around this handling. In that light, I would err on the sake of simplicity, and only add checks when we know we need them.

In this case, by having a singular merging behavior — allowing it regardless of props set — we'd also never run into the question of "why did it work this one time, but not that other"?

Not a strong opinion, but it's one that makes me lean towards starting with less.

Reply:
For me that's fine if we're ok with the potential content loss. E.g. if we're merging two quotes with cites, the cite of the quote we're merging will be lost. Is that ok? I added these checks because I though that's not ok.